### PR TITLE
Move delete comment authorization to pundit

### PR DESCRIPTION
**Pull request by DavidKang**

### CODE REVIEW

File: src/api/app/policies/comment_policy.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+class CommentPolicy < ApplicationPolicy
+  def initialize(user, record)
```
Comment (Ana06) [inline src/api/app/policies/comment_policy.rb]: I would say that you don't need that method, as it is in` application_policy.rb.` The only difference will be that if the user is not logged in it will raise the exception there instead of in the comment controller. But in my opinion this is even better as the error will be _must be logged in_ instead of _No permissions to delete comment_

File: src/api/app/policies/comment_policy.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    raise Pundit::NotAuthorizedError, "record does not exist" unless record
+    @user = user
+    @record = record
+  end
+
+  def destroy?
+    return false if @user.blank?
```
Comment (Ana06) [inline src/api/app/policies/comment_policy.rb]: if the initialize method is removed this is not needed :wink:

File: src/api/app/views/webui/comment/_links.html.erb [hunk 3]
```diff
@@ -6,11 +6,12 @@
     </a>
    </li>
   <% end %>
-  <% if comment.check_delete_permissions %>
+  <% if policy(comment).destroy? %>
       <li>
         <%= link_to(sprited_text('comment_delete', 'Delete'),
-                    {controller: :comments, action: :destroy,
-                     id: comment.id}, method: :delete) %>
+                    {controller: :comments, action: :destroy, id: comment.id},
```
Comment (Ana06) [inline src/api/app/views/webui/comment/_links.html.erb]: Why did you change these lines?

File: src/api/spec/policies/comment_policy_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+RSpec.describe CommentPolicy do
+  subject { CommentPolicy }
+  let(:comment_author) { create(:confirmed_user, login: 'burdenski') }
+  let(:admin_user) { create(:admin_user, login: 'admin') }
+  let(:user) { create(:confirmed_user, login: 'tom') }
+  let(:project) { create(:project, name: 'CommentableProject') }
+  let(:comment) { create(:comment_project, commentable: project, user: comment_author) }
+
+  permissions :destroy? do
+    it 'Users can destroy his own comments' do
```
Comment (Ana06) [inline src/api/spec/policies/comment_policy_spec.rb]: _Users can destroy **their** own comments_

File: src/api/spec/policies/comment_policy_spec.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      expect(subject).to permit(comment_author, comment)
+    end
+
+    it 'Other users cannot destroy comments from others' do
+      expect(subject).not_to permit(user, comment)
+    end
+
+    it 'Admin can destroy comments from other users' do
```
Comment (Ana06) [inline src/api/spec/policies/comment_policy_spec.rb]: _Admin can destroy comments from other **user**_

File: src/api/app/views/webui/comment/_links.html.erb (hunk 3 above)
Comment (DavidKang) [inline src/api/app/views/webui/comment/_links.html.erb]: Because you can delete a comment accidentally, so I added a confirmation to prevent this kind of situation. Furthermore, in most of the cases, I think that a delete action needs a double check.

File: src/api/app/policies/comment_policy.rb (hunk 1 above)
Comment (DavidKang) [inline src/api/app/policies/comment_policy.rb]: The reason that I'm overwriting this method is because we are using pundit helper in the views and  as the initialize method of `ApplicationPolicy` have `raise Pundit::NotAuthorizedError, "must be logged in" unless user` We couldn't realize any action If we are not logged, including see comments of a package

Comment (ChrisBr) [inline src/api/app/policies/comment_policy.rb]: Yep, I already discussed this a few weeks ago with @hennevogel. I think the solution from @DavidKang is fine, otherwise you always need to check in the view ``` if User.current && policy(comment).destroy?```. We can discuss maybe in a future PR if we should handle not logged in users different in the ApplicationPolicy, otherwise it's getting really ugly when we want to use the view helper more often.
Long story, short: For me this is fine for now!

File: src/api/app/views/webui/comment/_links.html.erb (hunk 3 above)
Comment (ChrisBr) [inline src/api/app/views/webui/comment/_links.html.erb]: I think this is also fine!

**Comment by ChrisBr** [comment]
LGTM :+1:
